lib/twocohom.gi

```diff
@@ -1590,20 +1697,221 @@ local r,z,ogens,n,gens,str,dim,i,j,f,rels,new,quot,g,p,lay,m,e,fp,old,sim,
+                  tab:=CopiedAugmentedCosetTable(tab0);
+                  tab.primaryImages:=Immutable(
+                    List(GeneratorsOfGroup(newfree){[1..nn]},
+                      x->ImagesRepresentative(e,x)));
+                  TrySecondaryImages(tab);
+
+                  i:=GroupHomomorphismByImagesNC(sub,Range(e),mwrd,
+                    List(GeneratorsOfGroup(newfree){[1..nn]},
+                      x->ImagesRepresentative(e,x)):noassert);
+                  SetCosetTableFpHom(i,tab);
+                  e:=PreImage(i,TrivialSubgroup(Range(e)));
+                  e:=Intersection(e,
+                      KernelOfMultiplicativeGeneralMapping(quot));
+
+                  # this check is very cheap, in comparison. So just be safe
+                  Assert(0,ForAll(RelatorsOfFpGroup(fps),
+                    x->IsOne(MappedWord(x,FreeGeneratorsOfFpGroup(fps),
+                      GeneratorsOfGroup(e!.quot)))));
+
+                  if step=0 then
+                    i:=GroupHomomorphismByImagesNC(e!.quot,p,
+                      GeneratorsOfGroup(e!.quot),
+                      GeneratorsOfGroup(p));
+                    j:=PreImage(i,m);
+                    if AbelianInvariants(j)=AbelianInvariants(newfree) then
+                      step:=2;
+                      Info(InfoExtReps,2,"Small bit did good");
+                    else
+                      Info(InfoExtReps,2,"Need expensive version");
+                      e:=LargerQuotientBySubgroupAbelianization(mfpi,m:
+                        cheap:=false);
+                      e:=Intersection(e,
+                        KernelOfMultiplicativeGeneralMapping(quot));
+                    fi;
+                  fi;
+                  
+
+                  step:=step+1;
+                od;
+
+              fi;
+
+            else
+              e:=LargerQuotientBySubgroupAbelianization(quot,m);
+              if e<>fail then
+                e:=Intersection(e,KernelOfMultiplicativeGeneralMapping(quot));
+              fi;
+            fi;
+
+
+            if e<>fail then
+              # can we do better degree -- greedy block reduction?
+              nn:=e!.quot;
+              if IsTransitive(nn,MovedPoints(nn)) then
+                repeat
+                  ei:=ShallowCopy(RepresentativesMinimalBlocks(nn,
+                    MovedPoints(nn)));
+                  SortBy(ei,x->-Length(x)); # long ones first
+                  j:=false;
+                  i:=1;
+                  while i<=Length(ei) and j=false do
+                    str:=Stabilizer(nn,ei[i],OnSets);
+                    if Size(Core(nn,str))=1 then
+                      j:=ei[i];
+                    fi;
+                    i:=i+1;
+                  od;
+                  if j<>false then
+                    Info(InfoExtReps,4,"deg. improved by blocks ",Size(j));
+                    # action on blocks
+                    i:=ActionHomomorphism(nn,Orbit(nn,j,OnSets),OnSets);
```

Comment:
Observation: The code in this `if` statement is never executed by the Travis test suite.